network/builder/src/builder.rs

```diff
@@ -496,6 +485,49 @@ impl NetworkBuilder {
         self
     }
 
+    fn add_onchain_discovery(
+        &mut self,
+        libra_db: Arc<dyn DbReader>,
+        waypoint: Waypoint,
+    ) -> &mut Self {
+        let (network_tx, discovery_events) = self
+            .add_protocol_handler(onchain_discovery::network_interface::network_endpoint_config());
+
+        let onchain_config = OnchainDiscoveryBuilderConfig::new(
+            self.conn_mgr_reqs_tx()
+                .expect("ConnectivityManager must be installed"),
+            network_tx,
+            discovery_events,
+            self.network_context().clone(),
+            libra_db,
+            waypoint,
+            // TODO: Move into NetworkConfig
+            30, // Legacy hard-coded value.
+            // TODO: move into NetworkConfig
+            8, // Legacy hard-coded value.
+        );
+
+        self.onchain_discovery_cfg = Some(onchain_config);
+
+        self.build_onchain_discovery()
+    }
+
+    fn build_onchain_discovery(&mut self) -> &mut Self {
+        if let Some(config) = self.onchain_discovery_cfg.take() {
+            let onchain_discovery_builder = OnchainDiscoveryBuilder::build(config, &self.executor);
+            self.onchain_discovery = Some(onchain_discovery_builder);
+            debug!("Built OnchainDiscovery");
+        }
+        self.start_onchain_discovery()
+    }
+
+    fn start_onchain_discovery(&mut self) -> &mut Self {
```

Comment:
prefer inline, this pattern doesn't make sense. I think I saw this earlier... so apologies if you responded.

Reply:
this is one of the important discussion points and a primary reason to make the PR visible sooner than later.

Reply:
Clarify:  this is one of the important low-level discussion points.    I am prioritizing the top level structure at present and can do lower level cleanups once the shape of the end goal has converged.